# Internship_T2T

# Table of contents
1. [Introduction](#introduction)
2. [Usage](#Usage)



## Introduction <a name="introduction"></a>

### text prone to change
Over the years, several human reference genomes have been developed with the most recent being the T2T-CHM13 reference genome. It adds nearly 200 million base pairs of sequence and unlocks complex regions of the human genome. This study performs two analysis to gain more insight on the read mapping coverage and structural variations when T2T-CHM13 is used. This is done to examine whether T2T-CHM13 could replace the currently used reference genome, Genome Reference Consortium Human Build 38 (GRCh38). The first analysis shows that the read mapping improves in several ways when mapped to T2T-CHM13, especially in the complex regions of the genome. However, no medically relevant genes that were exclusively  low coverage in GRCh38 had a significant read mapping coverage increase in T2T-CHM13. The second analysis indicates that the T2T-CHM13 is biased towards deletions, instead of an bias towards insertions similar to the GRCh38 reference genome. In addition, the gene SPC25 contained a T2T-exclusive heterozygous deletion showing that T2T-CHM13 can provide new structural variant for patient diagnostics. 


## Usage <a name="Usage"></a>

Go to the [documentation folder](https://github.com/WoutPoelen/Internship_T2T/tree/main/documentation) to find how to perform the steps taken in the internship.

Go to the [requirements folder](https://github.com/WoutPoelen/Internship_T2T/tree/main/requirements) to see the python requirements for this internship.

Go to the [results folder](https://github.com/WoutPoelen/Internship_T2T/tree/main/results) to see the figures produced throughout the internship.

Go to the [scripts folder](https://github.com/WoutPoelen/Internship_T2T/tree/main/scripts) to see the scripts made during the internship.
